streams/src/test/java/org/apache/kafka/streams/integration/KTableKTableForeignKeyJoinIntegrationTest.java

```diff
@@ -27,673 +25,314 @@
-        final List<KeyValue<Integer, String>> result2 = IntegrationTestUtils.waitUntilMinKeyValueRecordsReceived(
-                CONSUMER_CONFIG,
-                OUTPUT,
-                expected2.size());
-        assertEquals(expected2, result2);
-
-
-        leftTableEvents = Arrays.asList(new KeyValue<>(1, 1.11f));
-        IntegrationTestUtils.produceKeyValuesSynchronously(LEFT_TABLE, leftTableEvents, LEFT_PROD_CONF, MOCK_TIME);
-
-        final List<KeyValue<Integer, String>> expected3 = new LinkedList<>();
-        expected3.add(new KeyValue<>(1, "value1=1.11,value2=10"));
-        final List<KeyValue<Integer, String>> result3 = IntegrationTestUtils.waitUntilMinKeyValueRecordsReceived(
-                CONSUMER_CONFIG,
-                OUTPUT,
-                expected3.size());
-        assertEquals(expected3, result3);
-
-        //Ensure the state stores have the correct values within:
-        final Set<KeyValue<Integer, String>> expMatResults = new HashSet<>();
-        expMatResults.add(new KeyValue<>(1, "value1=1.11,value2=10"));
-        validateQueryableStoresContainExpectedKeyValues(expMatResults, currentMethodName);
+    public void doJoinFromLeftThenDeleteLeftEntity() {
+        for (final Boolean leftJoin : new Boolean[] {false, true}) {
+            final Topology topology = getTopology(STREAMS_CONFIG, "store", leftJoin);
+            try (final TopologyTestDriver driver = new TopologyTestDriver(topology, STREAMS_CONFIG)) {
+                final TestInputTopic<String, Long> right = driver.createInputTopic(RIGHT_TABLE, new StringSerializer(), new LongSerializer());
+                final TestInputTopic<Integer, Float> left = driver.createInputTopic(LEFT_TABLE, new IntegerSerializer(), new FloatSerializer());
+                final TestOutputTopic<Integer, String> outputTopic = driver.createOutputTopic(OUTPUT, new IntegerDeserializer(), new StringDeserializer());
+                final KeyValueStore<Integer, String> store = driver.getKeyValueStore("store");
+
+                right.pipeInput("1", 10L);
+                right.pipeInput("2", 20L);
+
+                left.pipeInput(1, 1.33f);
+                left.pipeInput(2, 2.77f);
+
+                {
+                    final Map<Integer, String> expected = mkMap(
+                        mkEntry(1, "value1=1.33,value2=10"),
+                        mkEntry(2, "value1=2.77,value2=20")
+                    );
+                    assertThat(
+                        "leftJoin:" + leftJoin,
+                        outputTopic.readKeyValuesToMap(),
+                        is(expected)
+                    );
+                    assertThat(
+                        "leftJoin:" + leftJoin,
+                        asMap(store),
+                        is(expected)
+                    );
+                }
+
+                //Now delete one LHS entity such that one delete is propagated down to the output.
+
+                left.pipeInput(1, null);
```

Comment:
I think we should delete each record and verify directly after a single delete

Reply:
Didn't follow this. You're saying we should delete all the records instead of just one? Would that result in additional coverage?

Reply:
I think so, especially if we delete a record that did not join and a record that does non-existing in the left table at all, because it should not emit a tombstone.

Reply:
These are good cases, but I still hesitate to pile too much onto these scenarios (it just makes them hard to follow). I'm adding new tests to cover these cases. Is this ok?